### ISSUE TITLE
Generating `:and` constrained recursive collections might get stuck

### ISSUE THREAD

**Issue by helins** [metosin/malli#442]
Generation enters some kind of infinite loop instead of stopping after a 100 trials as usually.

This happens in that kind of example:

```clojure
(malli.gen/generate [:and
                           {:registry {::data    [:or
                                                  ::int
                                                  ::vector]
                                       ::int     :int
                                       ::vector  [:vector
                                                  [:ref ::data]]}}
                           [:or ::vector]
                           [:fn
                            (fn [_]
                              false)]]
                          {:size 5})
```

Three conditions must present, other variations do not cause this problem:

- `:fn` in `:and` throws or returns false
- A recursive schema is involved
- This recursive schema is wrapped in something (`:or` like in this example, `:tuple`, `:cat`, ...)

Equally weird and intriguing.

**Comment by ikitommi** [comment]
weird. this still works:

```clj
(mg/generate
  [:and {:registry {::cons [:tuple :int [:maybe [:ref ::cons]]]
                    ::false [:fn (constantly false)]}}
   ::cons ::false]
  {:size 5})
; =throws=> Couldn't satisfy such-that predicate after 100 tries.
```

**Comment by helins** [comment]
Yes, I don't really know where to start. One clue is that is absolutely about generating a recursive item. If you replace `[:or ::vector]` with `[:or ::int]]` in my example, it fails as usually.

That way of defining recursive data is fairly straightforward to me, I don't think there is anything wrong with it? You could easily define JSON like that, for instance.

**Comment by ikitommi** [comment]
the definition is ok, there is a bug in the ref-generator lazyness. Need to debug.

**Comment by helins** [comment]
Just had an intuition. I am wondering if we haven't been blinded by the infamous halting problem.

Malli suffers from exponential explosion when generating recursive structures. I'll think a bit more about that and open an issue. `test.check` already provides a pretty good solution, it is a matter of how to integrate that into Malli.

Meanwhile, could you please check that the given **size** is always respected? Looking at the source, it looks like it is, but you might see something I don't. We are looking for a place where, somehow when retrying, the given size is not respected and the default size of 30 is used instead.

With a size of 30, generated structures from my example tend to be absolutely huge (that "exponential explosion" problem). Indeed, if I let my example run for a while (5-10 minutes), I get:

```clojure
; Execution error (OutOfMemoryError) at clojure.test.check.generators/halfs (generators.cljc:207).
; Java heap space
```

What do you think?

**Comment by opqdonut** [comment]
In current malli, the first example fails with `:malli.generator/such-that-failure`. See #1144 .
